File: mdnn/mdnn_cpp/src/atomic_nn/atomic_nn.cpp
#include "head.h"

#include "atomic_nn/atomic_nn.h"

AtomicNetImpl::AtomicNetImpl(int input_nodes, vector<int> hidden_nodes){
    Linear layer = nullptr;
    for (int i = 0; i < hidden_nodes.size(); i++){
        if (i == 0){
            layer = Linear(input_nodes, hidden_nodes[i]);
        }
        else{
            layer = Linear(hidden_nodes[i - 1], hidden_nodes[i]);
        }
        register_module("layer" + to_string(i), layer);
        layers.push_back(layer);
    }
    layer = Linear(hidden_nodes[hidden_nodes.size() - 1], 1);
    register_module("layer" + to_string(hidden_nodes.size()), layer);
    layers.push_back(layer); 
}

Tensor AtomicNetImpl::forward(Tensor g)
{
    Tensor x = g;
    for (int i = 0; i < layers.size() - 1; i++){
        x = layers[i]->forward(x);
        x = torch::sigmoid(x);
    }
    x = layers[layers.size() - 1]->forward(x);
    return x;
}

File: nnmd/nn/atomic_nn.py
import torch
import torch.nn as nn


class AtomicNN(nn.Module):
    """Model implements a multilayer perceptron
    for one species with a single output for one atom.
    """

    def __init__(
        self,
        input_size: int,
        hidden_sizes: list[int] | int,
        output_size: int = 1,
        activation: str = torch.sigmoid,
    ):
        """Create an instance of AtomicNN

        Args:
            input_size (int): inputs amount of NN
            hidden_sizes (list[int] | int): configuration of hidden layers nodes
            output_size (int): outputs amount of NN
            activation (str): activation function to use in hidden layers
        """
        super().__init__()

        self.model = nn.Sequential()
        if isinstance(hidden_sizes, list):
            self.model.append(nn.Linear(input_size, hidden_sizes[0]))
            for prev, curr in zip(
                range(len(hidden_sizes) - 1), range(1, len(hidden_sizes))
            ):
                self.model.append(nn.Linear(hidden_sizes[prev], hidden_sizes[curr]))
            self.model.append(nn.Linear(hidden_sizes[-1], output_size))
        elif isinstance(hidden_sizes, int):
            self.model.append(nn.Linear(input_size, hidden_sizes))
            self.model.append(nn.Linear(hidden_sizes, output_size))
        else:
            raise ValueError("hidden_sizes must be a list or an integer")

        self.activation = activation

    def forward(self, x):
        for i, layer in enumerate(self.model):
            x = layer(x)
            if i != len(self.model) - 1:
                x = self.activation(x)
        return x
